conversion: Return typed error on failure to convert for CRs

MadhavJivrajani opened the pull request:
<!--  Thanks for sending a pull request!  Here are some tips for you:

1. If this is your first time, please read our contributor guidelines: https://git.k8s.io/community/contributors/guide/first-contribution.md#your-first-contribution and developer guide https://git.k8s.io/community/contributors/devel/development.md#development-guide
2. Please label this pull request according to what type of issue you are addressing, especially if this is a release targeted pull request. For reference on required PR/issue labels, read here:
https://git.k8s.io/community/contributors/devel/sig-release/release.md#issuepr-kind-label
3. Ensure you have added or ran the appropriate tests for your PR: https://git.k8s.io/community/contributors/devel/sig-testing/testing.md
4. If you want *faster* PR reviews, read how: https://git.k8s.io/community/contributors/guide/pull-requests.md#best-practices-for-faster-reviews
5. If the PR is unfinished, see how to mark it: https://git.k8s.io/community/contributors/guide/pull-requests.md#marking-unfinished-pull-requests
-->

#### What type of PR is this?

/kind bug
<!--
Add one of the following kinds:
/kind bug
/kind cleanup
/kind documentation
/kind feature

Optionally add one or more of the following kinds if applicable:
/kind api-change
/kind deprecation
/kind failing-test
/kind flake
/kind regression
-->

#### What this PR does / why we need it:

Return a typed error that also passes as a "Missing Version Error"
so that server side apply updates are not blocked from happening on
custom resources that have a version that is no longer available.

This commit also adds an integration test to verify the same.

#### Which issue(s) this PR fixes:
<!--
*Automatically closes linked issue when PR is merged.
Usage: `Fixes #<issue number>`, or `Fixes (paste link of issue)`.
_If PR is about `failing-tests or flakes`, please post the related issues/tests in a comment and do not use `Fixes`_*
-->
Fixes https://github.com/kubernetes/kubernetes/issues/111937

#### Special notes for your reviewer:

#### Does this PR introduce a user-facing change?
<!--
If no, just write "NONE" in the release-note block below.
If yes, a release note is required:
Enter your extended release note in the block below. If the PR requires additional action from users switching to the new release, include the string "action required".

For more information on release notes see: https://git.k8s.io/community/contributors/guide/release-notes.md
-->
```release-note
NONE
```

#### Additional documentation e.g., KEPs (Kubernetes Enhancement Proposals), usage docs, etc.:

<!--
This section can be blank if this pull request does not require a release note.

When adding links which point to resources within git repositories, like
KEPs or supporting documentation, please reference a specific commit and avoid
linking directly to the master branch. This ensures that links reference a
specific point in time, rather than a document that may change over time.

See here for guidance on getting permanent links to files: https://help.github.com/en/articles/getting-permanent-links-to-files

Please use the following format for linking documentation:
- [KEP]: <link>
- [Usage]: <link>
- [Other doc]: <link>
-->
```docs

```

apelisse reviewed (commented):
I'm not super familiar with this code but that seems reasonable. Can you add a test to reproduce the bug that this is attempting to fix in ssa? Thank you!

MadhavJivrajani commented:
@apelisse I've added a reproducer test for the bug now, could you PTAL? Thanks!

MadhavJivrajani commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:
```diff
@@ -186,8 +186,9 @@ func (c *delegatingCRConverter) ConvertToVersion(in runtime.Object, target runti
-		return nil, fmt.Errorf("%v is unstructured and is not suitable for converting to %q", fromGVK.String(), target)
+		// // TODO: should this be a typed error?
+		// return nil, fmt.Errorf("%v is unstructured and is not suitable for converting to %q", fromGVK.String(), target)
+		return nil, runtime.NewNotRegisteredGVKErrForTarget("conversion.delegatingCRConverter", fromGVK, target)
```
There's an instance in the apiextentions-apiserver where this is used, I put the scheme name based on how its used there: 

https://github.com/kubernetes/kubernetes/blob/7e40d6d04ef8078772d2fa810b1ba925c9d8ace7/staging/src/k8s.io/apiextensions-apiserver/pkg/crdserverscheme/unstructured.go#L58

liggitt replied:
Normally, NewNotRegisteredGVKErrForTarget means there was a programmer error registering types into a runtime.Scheme, so the scheme name identifies the location where the runtime.Scheme instance was constructed so the programmer can find and fix the scheme instance.

It's going to be confusing to return this text about a CRD to an end user:

https://github.com/kubernetes/kubernetes/blob/778b24c97e189dcc8f14a81ce32e022b959f8fd3/staging/src/k8s.io/apimachinery/pkg/runtime/error.go#L50-L69

Can you exercise each of these paths in a unit test that surfaces the actual error text a user will see, so we can ensure that is understandable and sensible? We may want to make an alternate constructor for these error types that lets us set alternate context info that is not oriented around runtime.Scheme

Possibly something like this:

```diff
diff --git a/staging/src/k8s.io/apimachinery/pkg/runtime/error.go b/staging/src/k8s.io/apimachinery/pkg/runtime/error.go
index 7dfa45762fa..fbc00ff88ac 100644
--- a/staging/src/k8s.io/apimachinery/pkg/runtime/error.go
+++ b/staging/src/k8s.io/apimachinery/pkg/runtime/error.go
@@ -19,53 +19,76 @@ package runtime
 import (
 	"fmt"
 	"reflect"
+	"strconv"
 	"strings"
 
 	"k8s.io/apimachinery/pkg/runtime/schema"
 )
 
 type notRegisteredErr struct {
-	schemeName string
-	gvk        schema.GroupVersionKind
-	target     GroupVersioner
-	t          reflect.Type
+	context fmt.Stringer
+	gvk     schema.GroupVersionKind
+	target  GroupVersioner
+	t       reflect.Type
 }
 
 func NewNotRegisteredErrForKind(schemeName string, gvk schema.GroupVersionKind) error {
-	return &notRegisteredErr{schemeName: schemeName, gvk: gvk}
+	return &notRegisteredErr{context: schemeContext(schemeName), gvk: gvk}
 }
 
 func NewNotRegisteredErrForType(schemeName string, t reflect.Type) error {
-	return &notRegisteredErr{schemeName: schemeName, t: t}
+	return &notRegisteredErr{context: schemeContext(schemeName), t: t}
 }
 
 func NewNotRegisteredErrForTarget(schemeName string, t reflect.Type, target GroupVersioner) error {
-	return &notRegisteredErr{schemeName: schemeName, t: t, target: target}
+	return &notRegisteredErr{context: schemeContext(schemeName), t: t, target: target}
 }
 
 func NewNotRegisteredGVKErrForTarget(schemeName string, gvk schema.GroupVersionKind, target GroupVersioner) error {
-	return &notRegisteredErr{schemeName: schemeName, gvk: gvk, target: target}
+	return &notRegisteredErr{context: schemeContext(schemeName), gvk: gvk, target: target}
+}
+
+func NewNotRegisteredGVKErrForTargetWithContext(context string, gvk schema.GroupVersionKind, target GroupVersioner) error {
+	return &notRegisteredErr{context: genericContext(context), gvk: gvk, target: target}
+}
+
+type schemeContext string
+
+func (s schemeContext) String() string {
+	if len(s) == 0 {
+		return ""
+	}
+	return " in scheme " + strconv.Quote(string(s))
+}
+
+type genericContext string
+
+func (s genericContext) String() string {
+	if len(s) == 0 {
+		return ""
+	}
+	return " " + string(s)
 }
 
 func (k *notRegisteredErr) Error() string {
 	if k.t != nil && k.target != nil {
-		return fmt.Sprintf("%v is not suitable for converting to %q in scheme %q", k.t, k.target, k.schemeName)
+		return fmt.Sprintf("%v is not suitable for converting to %q%s", k.t, k.target, k.context)
 	}
 	nullGVK := schema.GroupVersionKind{}
 	if k.gvk != nullGVK && k.target != nil {
-		return fmt.Sprintf("%q is not suitable for converting to %q in scheme %q", k.gvk.GroupVersion(), k.target, k.schemeName)
+		return fmt.Sprintf("%q is not suitable for converting to %q%s", k.gvk.GroupVersion(), k.target, k.context)
 	}
 	if k.t != nil {
-		return fmt.Sprintf("no kind is registered for the type %v in scheme %q", k.t, k.schemeName)
+		return fmt.Sprintf("no kind is registered for the type %v%s", k.t, k.context)
 	}
 	if len(k.gvk.Kind) == 0 {
-		return fmt.Sprintf("no version %q has been registered in scheme %q", k.gvk.GroupVersion(), k.schemeName)
+		return fmt.Sprintf("no version %q has been registered%s", k.gvk.GroupVersion(), k.context)
 	}
 	if k.gvk.Version == APIVersionInternal {
-		return fmt.Sprintf("no kind %q is registered for the internal version of group %q in scheme %q", k.gvk.Kind, k.gvk.Group, k.schemeName)
+		return fmt.Sprintf("no kind %q is registered for the internal version of group %q%s", k.gvk.Kind, k.gvk.Group, k.context)
 	}
 
-	return fmt.Sprintf("no kind %q is registered for version %q in scheme %q", k.gvk.Kind, k.gvk.GroupVersion(), k.schemeName)
+	return fmt.Sprintf("no kind %q is registered for version %q%s", k.gvk.Kind, k.gvk.GroupVersion(), k.context)
 }
 
 // IsNotRegisteredError returns true if the error indicates the provided
```

MadhavJivrajani commented:
/test pull-kubernetes-conformance-kind-ipv6-parallel
(container not ready)

MadhavJivrajani commented:
/kind bug

cici37 commented:
/triage accepted

MadhavJivrajani commented:
/retest

MadhavJivrajani commented:
@liggitt thanks for the review, I've tried adding what you suggested, could you PTAL and see if this makes things better?

apelisse commented on test/integration/apiserver/apply/apply_crd_test.go:
```diff
@@ -733,3 +735,206 @@ spec:
+		Do(context.TODO())
+
+	resp, _ := new.Raw()
+	if new.Error() != nil && strings.Contains(new.Error().Error(), "request to convert CR to an invalid group/version") {
```
You should probably fail even if the error isn't the one you expected, no?

apelisse commented on test/integration/apiserver/apply/apply_crd_test.go:
```diff
@@ -733,3 +735,206 @@ spec:
+// in a CR: if a CR exists with a managedFields entry whose API Version is a version
+// that no longer exists, if a SSA operation is now done on this CR, the API Server
+// tries to perform conversion and fails. This should *not* hinder workflows and
+// updates should still be possible.
```
So the logic of SSA here is that it will remove the entry of the wrong version specifically, but not the others. So I would suggest that maybe the test includes:
- Have entries in old and new versino
- Remove old version
- Make sure entries initially written in new version persist (iow, make sure we haven't deleted all entries)

MadhavJivrajani replied:
I'm not sure I understand
> make sure we haven't deleted all entries

Do you mean in the end we check that `managedFields` is not empty?

apelisse replied:
I don't know if not empty is strong enough, but we need to make sure this doesn't delete entries that shouldn't be deleted.

liggitt commented on staging/src/k8s.io/apimachinery/pkg/runtime/error_test.go:
```diff
@@ -0,0 +1,106 @@
+		{
+			name:           "with type and target",
+			err:            NewNotRegisteredErrForTarget(schemeName, testType, target),
+			expectedErrStr: fmt.Sprintf("%v is not suitable for converting to %q%s", testType, target, context),
```
make `expectedErrStr` the literal string with no formatting so we can see what the final string actually looks like... that would have caught the missing space in schemeContext#String()

liggitt commented on staging/src/k8s.io/apimachinery/pkg/runtime/error.go:
```diff
@@ -19,53 +19,76 @@ package runtime
+	if len(s) == 0 {
+		return ""
+	}
+	return string(s)
```
```suggestion
	return " " + string(s)
```

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:
```diff
@@ -296,12 +295,13 @@ func getObjectsToConvert(
-			return nil, fmt.Errorf("request to convert CR list failed, list index %d has invalid group/version: %s", i, expectedGV.String())
+		expectedGVK := list.Items[i].GroupVersionKind()
+		if !validVersions[expectedGVK.GroupVersion()] {
+			return nil, runtime.NewNotRegisteredGVKErrForTargetWithContext("", expectedGVK, target)
```
```suggestion
			return nil, runtime.NewNotRegisteredGVKErrForTargetWithContext(fmt.Sprintf("in list item %d", i), expectedGVK, target)
```

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:
```diff
@@ -204,17 +203,17 @@ func (c *delegatingCRConverter) ConvertToVersion(in runtime.Object, target runti
 	desiredAPIVersion := toGVK.GroupVersion().String()
 	if !c.validVersions[toGVK.GroupVersion()] {
-		return nil, fmt.Errorf("request to convert CR to an invalid group/version: %s", desiredAPIVersion)
+		return nil, runtime.NewNotRegisteredGVKErrForTargetWithContext("", toGVK, target)
```
this isn't going to make sense... it will say "$group/$version is not suitable for converting to $group/version", right? is there a test that exercises this code path where we can see the message that gets produced?

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:
```diff
@@ -204,17 +203,17 @@ func (c *delegatingCRConverter) ConvertToVersion(in runtime.Object, target runti
 
 	var objectsToConvert []*unstructured.Unstructured
-	objectsToConvert, err := getObjectsToConvert(list, desiredAPIVersion, c.validVersions)
+	objectsToConvert, err := getObjectsToConvert(list, desiredAPIVersion, c.validVersions, target)
```
```suggestion
	objectsToConvert, err := getObjectsToConvert(list, desiredAPIVersion, c.validVersions, toGVK.GroupVersion())
```

liggitt commented on staging/src/k8s.io/apiextensions-apiserver/pkg/apiserver/conversion/converter.go:
```diff
@@ -204,17 +203,17 @@ func (c *delegatingCRConverter) ConvertToVersion(in runtime.Object, target runti
 	// get a v1.List.
 	if !c.validVersions[fromGVK.GroupVersion()] {
-		return nil, fmt.Errorf("request to convert CR from an invalid group/version: %s", fromGVK.GroupVersion().String())
+		return nil, runtime.NewNotRegisteredGVKErrForTargetWithContext("", fromGVK, target)
```
```suggestion
		return nil, runtime.NewNotRegisteredGVKErrForTargetWithContext("", fromGVK, toGVK.GroupVersion())
```

MadhavJivrajani commented:
Thanks for the reviews! Planning to get back to this early next week.

MadhavJivrajani committed:
runtime: add ability to provide custom strings instead of scheme name

This commit adds the ability to inject a custom string instead of the
default scheme name. It also consequently edits what is printed when
a custom string is passed.

This commit also adds tests to exercise the different code paths taken
for notRegisteredErr.

This is needed when we want to return typed errors and scheme names are
not known before hand, such as conversion in custom resources.

Signed-off-by: Madhav Jivrajani <madhav.jiv@gmail.com>

MadhavJivrajani committed:
conversion: return typed error on failure to convert

Return a typed error that also passes as a "Missing Version Error"
so that server side apply updates are not blocked from happening on
custom resources that have a version that is no longer available.

This commit also adds an integration test to verify the same.

Signed-off-by: Madhav Jivrajani <madhav.jiv@gmail.com>

MadhavJivrajani commented on staging/src/k8s.io/apimachinery/pkg/runtime/error.go:
```diff
@@ -19,53 +19,76 @@ package runtime
+	return &notRegisteredErr{context: schemeContext(schemeName), gvk: gvk, target: target}
+}
+
+func NewNotRegisteredGVKErrForTargetWithContext(context string, gvk schema.GroupVersionKind, target GroupVersioner) error {
```
Yeah this was due to my incorrect usage of this, changed it.

MadhavJivrajani commented:
@liggitt @apelisse I've addressed most comments (except https://github.com/kubernetes/kubernetes/pull/115650#discussion_r1178502155), could you PTAL? 
Apologies, this slipped through the cracks 🙏🏼

apelisse commented on test/integration/apiserver/apply/apply_crd_test.go:
```diff
@@ -733,3 +734,206 @@ spec:
+	// Migrate existing CR to new storage version by applying an
+	// empty merge patch.
+	_, err = dynamicClient.Resource(gvr).
+		Patch(context.TODO(), name, types.MergePatchType, []byte("{}"), metav1.PatchOptions{})
```
Maybe send an apply in the new version that changes other fields, so that you have two entries at that point, one in the old version and one in the new version.

apelisse commented on test/integration/apiserver/apply/apply_crd_test.go:
```diff
@@ -733,3 +734,206 @@ spec:
+
+	apiVersion = noxuDefinition.Spec.Group + "/" + noxuDefinition.Spec.Versions[0].Name
+	yamlBody = []byte(fmt.Sprintf(`
+apiVersion: %s
```
Make sure you send new fields in that apply so that you can see from the list of managed fields exactly what happened. (i.e. the old entry was removed even though it still should have owned some fields), the second entry that was inserted also still exists, and that last one also exist. You should have two entries at the end.

dprotaso commented:
@MadhavJivrajani bump

MadhavJivrajani commented:
Thanks for the ping, this had dropped off my radar. Planning to get back to this.

MadhavJivrajani committed:
flesh out integration test

Signed-off-by: Madhav Jivrajani <madhav.jiv@gmail.com>

MadhavJivrajani commented:
@apelisse 👋🏼 
I think I understand what you were asking from the test, but still not a 100% sure. 
I've pushed a commit with some log lines that tell what is happening at each stage, could you please take a look and correct me if I'm misunderstanding something here?

Copy pasting the output here for easy access:
```
    apply_crd_test.go:866: Created CR: {"apiVersion":"mygroup.example.com/v1beta2","kind":"WishIHadChosenNoxu","metadata":{"creationTimestamp":"2024-01-10T09:37:35Z","generation":1,"managedFields":[{"apiVersion":"mygroup.example.com/v1beta2","fieldsType":"FieldsV1","fieldsV1":{"f:spec":{"f:cronSpec":{},"f:ports":{"k:{\"containerPort\":80,\"protocol\":\"TCP\"}":{".":{},"f:containerPort":{},"f:name":{},"f:protocol":{}}},"f:replicas":{}}},"manager":"apply_test","operation":"Apply","time":"2024-01-10T09:37:35Z"}],"name":"mytest","resourceVersion":"90","uid":"e306af17-e9e1-4efb-b480-49b79da8fd9f"},"spec":{"cronSpec":"* * * * */5","ports":[{"containerPort":80,"name":"x","protocol":"TCP"}],"replicas":1}}
        
    apply_crd_test.go:882: 
        
        Updated CRD, migrated away from old storage version: [v1beta2 v1beta1]
    apply_crd_test.go:898: 
        
        Migrated CR to new storage version after patch (should have 2 managedFields entries in different versions): &{map[apiVersion:mygroup.example.com/v1beta1 kind:WishIHadChosenNoxu metadata:map[creationTimestamp:2024-01-10T09:37:35Z generation:2 managedFields:[map[apiVersion:mygroup.example.com/v1beta2 fieldsType:FieldsV1 fieldsV1:map[f:spec:map[f:cronSpec:map[] f:ports:map[k:{"containerPort":80,"protocol":"TCP"}:map[.:map[] f:containerPort:map[] f:name:map[] f:protocol:map[]]]]] manager:apply_test operation:Apply time:2024-01-10T09:37:35Z] map[apiVersion:mygroup.example.com/v1beta1 fieldsType:FieldsV1 fieldsV1:map[f:spec:map[f:replicas:map[]]] manager:apply.test operation:Update time:2024-01-10T09:37:35Z]] name:mytest resourceVersion:92 uid:e306af17-e9e1-4efb-b480-49b79da8fd9f] spec:map[cronSpec:* * * * */5 ports:[map[containerPort:80 name:x protocol:TCP]] replicas:2]]}
    apply_crd_test.go:911: 
        
        Updated CRD, remove old storage version from status: [v1beta1]
    apply_crd_test.go:954: 
        
        applying object again but with changed replicas (should have 2 converted managedFields entries): {"apiVersion":"mygroup.example.com/v1beta1","kind":"WishIHadChosenNoxu","metadata":{"creationTimestamp":"2024-01-10T09:37:35Z","generation":2,"managedFields":[{"apiVersion":"mygroup.example.com/v1beta1","fieldsType":"FieldsV1","fieldsV1":{"f:spec":{"f:cronSpec":{},"f:ports":{"k:{\"containerPort\":80,\"protocol\":\"TCP\"}":{".":{},"f:containerPort":{},"f:name":{},"f:protocol":{}}},"f:replicas":{}}},"manager":"apply_test","operation":"Apply","time":"2024-01-10T09:37:35Z"},{"apiVersion":"mygroup.example.com/v1beta1","fieldsType":"FieldsV1","fieldsV1":{"f:spec":{"f:replicas":{}}},"manager":"apply.test","operation":"Update","time":"2024-01-10T09:37:35Z"}],"name":"mytest","resourceVersion":"96","uid":"e306af17-e9e1-4efb-b480-49b79da8fd9f"},"spec":{"cronSpec":"* * * * */5","ports":[{"containerPort":80,"name":"x","protocol":"TCP"}],"replicas":2}}
```

MadhavJivrajani commented:
If I run the added test against `master` that does not have the new typed error logic, I get a failure consistent with the original bug report:
```
 request to convert CR to an invalid group/version: mygroup.example.com/v1beta2
```

MadhavJivrajani commented:
/retest

MadhavJivrajani commented:
/retest

dprotaso commented:
Can we cherry pick this change back to supported releases? This issue has affected K8s since 1.22

jpbetz commented on staging/src/k8s.io/apimachinery/pkg/runtime/error.go:
```diff
@@ -19,53 +19,76 @@ package runtime
+	if len(s) == 0 {
+		return ""
+	}
+	return " in scheme " + strconv.Quote(string(s))
```
nit: Padding the context with a space prefix here feels clunky.  Could we do all the formatting in one place?

MadhavJivrajani replied:
@jpbetz the reason for this is in how the `notRegisteredErr` string is formatted here: https://github.com/MadhavJivrajani/kubernetes/blob/fa6dcc734bf6d6faa720714628b8709c27736f19/staging/src/k8s.io/apimachinery/pkg/runtime/error.go#L73

There are two main cases to take care of: when we use a `genericContext` and when we use a `schemeContext`. If the `genericContext` is used (meaning we already know what the scheme is, such as in cases of built in types), things would be okay for the most part, however, if we use a `schemeContext` (for when we don't know the scheme before hand such as in CRs), we need to add in a ` in scheme ` there towards the _end_ of the formatted `notRegisteredErr` error string. So to handle both cases and not overcomplicate the formatting logic, a space is padded here.

MadhavJivrajani commented on staging/src/k8s.io/apimachinery/pkg/runtime/error.go:
```diff
@@ -19,53 +19,76 @@ package runtime
+	return &notRegisteredErr{context: schemeContext(schemeName), gvk: gvk, target: target}
+}
+
+func NewNotRegisteredGVKErrForTargetWithContext(context string, gvk schema.GroupVersionKind, target GroupVersioner) error {
```
`context` is probably a misnomer, but happy to change what it is called.

> It makes it easy to bypass the use of genericContext (..and the formatting that genericContext adds, which makes me think we should handle the formatting differently..)

Not sure I understand the bypassing part here, could you please elaborate?

MadhavJivrajani replied:
cc @liggitt

fabriziopandini commented:
This issue is biting us pretty hard in Cluster API, is there any chance to backport this to older releases of Kubernetes (as much as possible?)

dims commented:
/assign @liggitt @apelisse

liggitt commented:
> This issue is biting us pretty hard in Cluster API, is there any chance to backport this to older releases of Kubernetes (as much as possible?)

It's *really* hard to reason about the implications of the change, so doing that safely on master is the first step. It's too early to comment about the possibility of backports, but my inclination would be not to backport, since this isn't a regression (and the implications are super hard to reason about).

dprotaso commented:
> but my inclination would be not to backport

Yeah sounds good @liggitt 

@MadhavJivrajani Is there any other action here for this PR?

fabriziopandini commented:
Sorry folks to ping again, but this issue is preventing projects to drop old API versions.
Any chances this will be fixed on main?

jpbetz commented:
> Sorry folks to ping again, but this issue is preventing projects to drop old API versions. Any chances this will be fixed on main?

https://github.com/kubernetes/kubernetes/pull/115650/files#r1253442734 appears to be outstanding

liggitt commented:
/remove-lifecycle rotten

dprotaso commented:
/reopen

dprotaso commented:
@MadhavJivrajani is this something you can wrap up?
